### PR TITLE
traceroute support for SNAT mode (ICMP method)

### PR DESCRIPTION
**Pull request by yiflv**

### CODE REVIEW

File: src/ipvs/ip_vs_proto_icmp.c [hunk 1]
```diff
@@ -224,8 +225,31 @@ static struct dp_vs_conn *icmp_conn_lookup(struct dp_vs_proto *proto,
...
+                                                        sizeof(_icmph)))
+                    return NULL;
+ 
+                s_addr.in.s_addr = ip_hdr->daddr;
+                /* get inner ICMP header */
+                ich = ip_hdr + 1;
+                icmp_type = ((struct icmphdr *)ich)->type;
+                icmp_code = ((struct icmphdr *)ich)->code;
+                sport = icmp_type << 8 | icmp_code;
+                dport = ((struct icmphdr *)ich)->un.echo.id;
```
Comment (ywc689) [inline src/ipvs/ip_vs_proto_icmp.c:248]: Does it work when packet redirect is not ON?

Comment (yiflv) [reply src/ipvs/ip_vs_proto_icmp.c:248]: > Does it work when packet redirect is not ON?

No, it needs packet redirect. So this feature does no good to performance.

File: src/ipvs/ip_vs_core.c [hunk 2]
```diff
@@ -704,6 +704,11 @@ static int __dp_vs_in_icmp4(struct rte_mbuf *mbuf, int *related)
         return INET_DROP;
     dp_vs_fill_iphdr(AF_INET, mbuf, &dciph);
 
+    if(ich->type == ICMP_TIME_EXCEEDED) {
```
[the remark this answers is not in the document]
Comment (donghaobo) [reply src/ipvs/ip_vs_core.c:707]: the linux kernel does not handle this situation either

File: src/ipv4.c [hunk 3]
```diff
@@ -428,6 +431,15 @@ static int ipv4_rcv(struct rte_mbuf *mbuf, struct netif_port *port)
     ip4_show_hdr(__func__, mbuf);
 #endif
 
+    if (unlikely(iph->time_to_live <= 1)) {
```
[the remark this answers is not in the document]
Comment (yiflv) [reply src/ipv4.c:434]: Yes, I knew it and I saw your pr. But it seems not a good idea to remove source check directly (refer to #475).

[the remark this answers is not in the document]
Comment (yiflv) [reply src/ipv4.c:434]: > ttl check is already done at xmit, enable by xmit_ttl
Yes, I knew it and I saw your pr. But it seems not a good idea to remove source check directly (refer to #475).

[the remark this answers is not in the document]
Comment (yiflv) [reply src/ipv4.c:434]: > ttl check is already done at xmit, enable by xmit_ttl

Yes, I knew it and I saw your pr. But it seems not a good idea to remove source check directly (refer to #475). @donghaobo

File: src/ipvs/ip_vs_core.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiflv) [reply src/ipvs/ip_vs_core.c:707]: Sorry, I didn't get your meaning. The reason SNAT failed is ICMP_TIME_EXCEEDED packets coming back failed to find its conn in the lcore it reached (tuple and hash changed). Other devices (routers) doesn't need to translate the inner ip. Only DPVS machine handle this job.

[the remark this answers is not in the document]
Comment (donghaobo) [reply src/ipvs/ip_vs_core.c:707]: > Other devices (routers) doesn't need to translate the inner ip. Only DPVS machine handle this job.

i don't agree, all NAT devices should translate inner ip, see rfc 3022  section 4.3.
yet, your change is not wrong.

File: src/ipv4.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (donghaobo) [reply src/ipv4.c:434]: > > ttl check is already done at xmit, enable by xmit_ttl
> 
> Yes, I knew it and I saw your pr. But it seems not a good idea to remove source check directly (refer to #475). @donghaobo

if than change is not good, we shold make it right, not leave it not working and inplement another @yifan

**Pull request closed without merging**